[2.9] various elbv2 bugfixes

felixfontein opened the pull request:
##### SUMMARY
Backport of #65021 and #65183 to stable-2.9.

CC @jillr @mjmayer @s-hertel @markuman

Could you please check whether the changelog fragment I wrote is appropriate? It looks like only the two modules mentioned use the `elbv2` module utils.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/module_utils/aws/elbv2.py

mjmayer committed:
Add aws elb multiple host header support (#65021)

* Add support for aws elbv2 multiple host header actions

* Add example of multiple host header

* Add test for multiple host headers

* Add missing quote

* Update test/integration/targets/elb_application_lb/tasks/test_modifying_alb_listeners.yml

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

* Update test/integration/targets/elb_application_lb/tasks/test_modifying_alb_listeners.yml

Co-Authored-By: Jill R <4121322+jillr@users.noreply.github.com>

* Remove debug statements

(cherry picked from commit d52af75c68b3f31c994d8b234b9c1e2387f9a4dd)

markuman committed:
elbv2 - print() statement results an task failure, even if the task itself was successfull (#65183)

(cherry picked from commit cba9abcbec6d9d11efea7b38f997e4ed16ba7687)

felixfontein committed:
Add changelog fragment.

markuman reviewed (approved):
LGTM

jillr reviewed (changes requested):
I'd like the changelog fragment to be a bit more specific, "various errors" and "some features" feels a bit mysterious and would leave me uncertain if I were a user. I'd rather we clearly state the problem was incorrectly formatted json and link to #65020.
Thanks for opening the backport!

felixfontein commented on changelogs/fragments/65021-65183-elb-bugs.yml:
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+- "elb_application_lb, elb_network_lb - fixed various Python errors in the shared module utils affecting some features of these modules."
```
@jillr how about:
```suggestion
- "elb_application_lb, elb_network_lb - 
fixed errors during listener rule comparison which caused broken JSON, and which caused some values not being compared correctly. (https://github.com/ansible/ansible/issues/65020)"
```

jillr replied:
That's great, thanks!

felixfontein committed:
Update changelogs/fragments/65021-65183-elb-bugs.yml

roberbri commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -717,7 +717,7 @@ def _compare_condition(self, current_conditions, condition):
                     condition_found = True
                     break
-            elif current_condition['Field'] == condition['Field'] and current_condition['Values'][0] == condition['Values'][0]:
+            elif current_condition['Field'] == condition['Field'] and current_condition['Values'] == condition['Values']:
```
Should the ['Values'] lists be sorted before the comparison?

mjmayer replied:
Good point, most of the time I anticipate `['Values']` will only be one item. But in #65021 it does support multiple values, which are a list of hostnames. I don't know if AWS returns that list in the same order all the time. Nevertheless, if someone made changes in the AWS console and then applied those changes through ansible, it is possible that the hostname order would be different and Ansible would report the configuration changed.

roberbri replied:
I have submitted #65315 for devel which should address this there. I'm not sure how to submit a change to someone elses PR.

felixfontein replied:
@roberbri changes need to be merged to `devel` before they can be backported, so you can't add something directly to this PR without going via `devel` first. Assuming #65315 gets merged before this PR gets merged, I can include a backport of it in this PR. If not, it can be added later with another backport PR.

nitzmahone commented:
I can't tell if this should be merged with the combo of multiple backports PRs and various seemingly unresolved discussion, so it's missing the bus for 2.9.2.

s-hertel commented on lib/ansible/module_utils/aws/elbv2.py:
```diff
@@ -717,7 +717,7 @@ def _compare_condition(self, current_conditions, condition):
                     condition_found = True
                     break
-            elif current_condition['Field'] == condition['Field'] and current_condition['Values'][0] == condition['Values'][0]:
+            elif current_condition['Field'] == condition['Field'] and current_condition['Values'] == condition['Values']:
```
+1 to include the fix in #65315 here. The rest of this PR looks good.

roberbri committed:
Force values to be sorted when compared. (#65315)

* Force values to be sorted when compared.

This avoids problems where lists of values could be generated in
indeterminate order.

* Update lib/ansible/module_utils/aws/elbv2.py

Thank you felixfontein, sorry I missed this detail.

Co-Authored-By: Felix Fontein <felix@fontein.de>
(cherry picked from commit 777b6d3e3098cbcf9f7e1f6d1b8f5fe04db5cf64)

felixfontein committed:
Update changelog.

felixfontein commented:
I've included #65315, and updated the changelog so it explicitly mentions the improved idempotence check. Could you please check if the wording is ok? Thanks :)

jillr commented:
LGTM, thanks @felixfontein

mattclay commented:
@felixfontein Thanks for the backport. In the future, if possible, please limit backports to a single PR. Combining multiple PRs into a single backport complicates the review process.
